smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -84,51 +145,182 @@ class DragonBackend:
-        print(f"| {self._updates}: Allocated hosts: ", self._allocated_hosts)
-        print(f"| {self._updates}: Running steps: ", self._running_steps)
-        print(f"| {self._updates}: Group infos: ", self._group_infos)
-        print(f"| {self._updates}: There are {len(self._queued_steps)} queued steps")
-        print("-------------------------------------------------------------\n")
+        """Whether the shutdown was requested to this server"""
+        self._can_shutdown = False
+        """Whether the server can shut down"""
+        self._frontend_shutdown: bool = False
+        """Whether the server frontend should shut down when the backend does"""
+        self._shutdown_initiation_time: t.Optional[float] = None
+        """The time at which the server initiated shutdown"""
+        smartsim_config = get_config()
+        self._cooldown_period = (
+            smartsim_config.telemetry_frequency * 2 + 5
+            if smartsim_config.telemetry_enabled
+            else 5
+        )
+        """Time in seconds needed to server to complete shutdown"""
+        logger.debug(f"{host_string} available for execution: {self._hosts}")
+
+    @staticmethod
+    def _proc_group_info_table_line(
+        step_id: str, proc_group_info: ProcessGroupInfo
+    ) -> t.List[str]:
+        table_line = [step_id, f"{str(proc_group_info.status)}"]
+
+        if proc_group_info.hosts is not None:
+            table_line.append(f"{','.join(proc_group_info.hosts)}")
+        else:
+            table_line.append("")
+
+        if proc_group_info.return_codes is not None:
+            table_line.append(
+                f"{','.join(str(ret) for ret in proc_group_info.return_codes)}"
+            )
+        else:
+            table_line.append("")
+
+        if proc_group_info.puids is not None:
+            table_line.append(f"{len(proc_group_info.puids)}")
+        else:
+            table_line.append("")
+
+        return table_line
 
     @property
-    def should_shutdown(self) -> bool:
-        return self._shutdown_requested
+    def step_table(self) -> str:
+        """Table representation of all jobs which have been started on the server."""
+        headers = ["Step", "Status", "Hosts", "Return codes", "Num procs"]
+        values = []
+
+        with self._queue_lock:
+            for step, group_info in self._group_infos.items():
+                values.append(
+                    DragonBackend._proc_group_info_table_line(step, group_info)
+                )
+
+        return tabulate(values, headers, disable_numparse=True, tablefmt="github")
+
+    @property
+    def host_table(self) -> str:
+        """Table representation of current state of nodes available
+
+        in the allocation.
+        """
+        headers = ["Host", "State"]
+        values = []
+
+        with self._queue_lock:
+            for host in self._hosts:
+                values.append([host, "Free" if host in self._free_hosts else "Busy"])
```

Comment:
It seems insane to lock something important to print this message. Can we rely on eventual consistency instead of introducing performance hits for string building?